File: client/src/models/Board.ts
import { Theme, ThemeProps } from "./Theme";
import { Tile, TileProps } from "./Tile";

export interface BoardProps {
    id: number;
    name: string;
    rows: number;
    cols: number;
    tiles?: TileProps[];
    themes?: ThemeProps[];
}

export class Board {
    public readonly id: number;
    public readonly name: string;
    public readonly rows: number;
    public readonly cols: number;
    public readonly tiles: Tile[];
    public readonly themes?: Theme[] | null;

    constructor(props: BoardProps) {
        this.id = props.id;
        this.name = props.name;
        this.rows = props.rows;
        this.cols = props.cols;
        this.tiles = props.tiles ? props.tiles.map((tile) => new Tile(tile)) : [];
        this.themes = props.themes ? props.themes.map((theme) => new Theme(theme)) : null;
    }
}

File: server/src/main/java/com/quest/models/Board.java
package com.quest.models;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;

@Entity
@Table(name = "boards")
public class Board {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "name", nullable = false)
    private String name;

    @Column(name = "row_count", nullable = false)
    private int rows;

    @Column(name = "col_count", nullable = false)
    private int cols;

    @JsonManagedReference
    @OneToMany(mappedBy = "board", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @OrderBy("sequence ASC")
    private List<Tile> tiles = new ArrayList<>();

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public int getRows() {
        return rows;
    }

    public void setRows(int rows) {
        this.rows = rows;
    }

    public int getCols() {
        return cols;
    }

    public void setCols(int cols) {
        this.cols = cols;
    }

    public List<Tile> getTiles() {
        return tiles;
    }

    public void setTiles(List<Tile> tiles) {
        this.tiles = tiles;
    }

    public Tile getStartTile() {
        if (tiles.isEmpty())
            throw new IllegalStateException("No tiles found in the board");
        return tiles.get(0);
    }

    public Optional<Tile> findTileById(Long id) {
        return tiles.stream()
                .filter(t -> t.getId().equals(id))
                .findFirst();
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }
}
